clients/src/main/java/org/apache/kafka/common/security/ssl/DefaultSslEngineFactory.java

```diff
@@ -255,7 +271,7 @@ private SSLContext createSSLContext(SecurityStore keystore, SecurityStore trusts
             }
 
             String tmfAlgorithm = this.tmfAlgorithm != null ? this.tmfAlgorithm : TrustManagerFactory.getDefaultAlgorithm();
-            TrustManagerFactory tmf = TrustManagerFactory.getInstance(tmfAlgorithm);
+            CommonNameLoggingTrustManagerFactoryWrapper tmf = CommonNameLoggingTrustManagerFactoryWrapper.getInstance(tmfAlgorithm);
```

Comment:
I think we should add a config to enable this logging. 

cc @rajinisivaram @ijuma

Reply:
I'm not sure if I can contribute to resolving this question. But here are some considerations.

Personally, I don't think that keeping this information/warning log message always enabled will cause any trouble in typical monitoring environments. But I'm not 100% sure about it and do not have any evidence. One thing to consider might be that the proposed changes are currently only affecting very few source files. Implementing a config option will make the changes more complex, further increase the number of affected parts of the code, and potentially require changes in user configurations (even though the default would be to disable this feature, I assume).

But of course, if you deem this necessary, I will implement it.

Reply:
My main concern is related to possible performance impact of the additional processing. We often see that faulty clients bombarding the brokers with handshake requests with expired certificates. This can impact broker performance. So I was thinking of adding a flag to enable/disable. Lets see what others thinking

Reply:
I have an alternative idea which might be worse discussing as it even speed up the handling of such faulty clients compared to the current implementation in trunk. We could cache negative decisions (i.e. rejected cert chains) and reject further client connections based on the cache without revalidating them. The idea is based on two assumptions:

Assumption 1: As long as the truststore is unmodified, any invalid certificate chain will never become valid, except if certificates are not yet valid (NotBefore constraint violated). The latter should rarely occur.
Assumption 2: Consulting an efficiently implemented decision cache would be significantly faster than validating a certificate chain.

The algorithm would be like this:
1. For any client certificate chain to be checked, we would first check if the truststore has been modified. 
2. If this is the case, we invalidate our cache. 
3. Otherwise, we calculate a hash over the whole cert chain and check if it is contained in our cache. 
4. If this is the case, we reject the chain. 
5. Otherwise we do the regular validation and accept the certificate if the validation is successful.
6. If the validation fails, we identify if this is due to the end certifcate being expired and log this.
7. We calculate the hash over the whole chain (see above) and add it to the cache, but only if this cert might not become valid in the future (NotBefore constraint violated).

The cache should be implemented using a hash set (lookup with ~ O(1)) which is bounded in size (to prevent attackers from causing OOM) by keeping only the latest N entries. The number N might be configurable.

Advantage: We speed up handling of all invalid certificates as soon as they have been found invalid once. Any valid certificate will just be affected by a very quick extra step (calc hash, lookup hash in reject cache).

What do you think about this approach?

Reply:
@omkreddy please have a look at the latest commits: I implemented a cache for faster rejecting known-invalid certificate chains. The test cases show that the result of the checks is equal to the unmodified trust store implementation, at least for the provided test data. However, a few extra cycles need to be spend on calculating the hash for the certificate chain and do a lookup in the cache, even if the certificate chain is valid (there is no free lunch).

Reply:
@rajinisivaram: I followed your advise,  added a new SSL factory implementation and removed my changes to the default SSL factory. The updated code can now be enabled by setting "ssl.engine.factory.class=org.apache.kafka.common.security.ssl.CommonNameLoggingSslEngineFactory" in the config file. However, as the default factory is "final" and hides a lot of its internal implementation details, there is some redundant code which can hardly be avoided without changing the default factory (e.g. removing "final", making some methods protected to allow inheritance, or adding some setters/getters to make it possible to delegate from the new implementation to the default one).